### PR TITLE
Add optional extent to metadata update failure logs

### PR DESCRIPTION
**Pull request by jschmidt10**
This week we saw many "Failed to do close consistency check" errors. The log showed that the accumulo.metadata table had additional RFiles that were not present in the tserver's DatafileManager.datafileSizes map. We believe that this may have been due to a temporary network drop where a bulk importing tserver sent accumulo.metadata updates - the metadata updates were persisted but the bulk importing tserver's writer client did not receive a successful response. 

We'd like to add some additional logging to MetadataTableUtil so that we can tie the failures back to particular tablets to see if these correlate.

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/data/Mutation.java [hunk 1]
```diff
@@ -1237,4 +1238,37 @@ protected SERIALIZED_FORMAT getSerializedFormat() {
...
+        sb.append(utf8String(update.getValue()));
+      }
+      sb.append('\n');
+    }
+
+    return sb.toString();
+  }
+
+  private static String utf8String(byte[] bytes) {
+    return new String(bytes, StandardCharsets.UTF_8);
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/data/Mutation.java:1272]: If you just do a static import of `StandardCharsets.UTF_8`, the lines where this is used become short enough that you don't need this separate private helper method.

File: core/src/main/java/org/apache/accumulo/core/data/Mutation.java [hunk 2]
```diff
@@ -1237,4 +1238,37 @@ protected SERIALIZED_FORMAT getSerializedFormat() {
...
 
+  /**
+   * Creates a multi-lined, human-readable String for this mutation.
+   *
+   * This method creates many intermediate Strings and should not be used for large volumes of
+   * Mutations.
+   *
+   * @return A multi-lined, human-readable String for this mutation.
+   */
+  public static String prettyPrint(Mutation mutation) {
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/data/Mutation.java:1249]: This is public API, so it would need a `@since` javadoc tag.
Also, why make this static?

Also, according to semantic versioning, we don't make API additions in bugfix releases, so this should not get added to 1.10, but could be added to 2.1, our next minor release planned.

File: server/base/src/main/java/org/apache/accumulo/server/master/balancer/HostRegexTableLoadBalancer.java [hunk 3]
```diff
@@ -34,7 +34,6 @@
 
 import org.apache.accumulo.core.client.admin.TableOperations;
 import org.apache.accumulo.core.conf.AccumuloConfiguration;
-import org.apache.accumulo.core.conf.ConfigurationObserver;
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/master/balancer/HostRegexTableLoadBalancer.java:37]: This seems unrelated. Is this an existing unused import?

File: server/base/src/main/java/org/apache/accumulo/server/util/MetadataTableUtil.java [hunk 4]
```diff
@@ -148,22 +148,24 @@ public static void update(ClientContext context, ZooLock zooLock, Mutation m, Ke
...
-        log.error("{}", e.getMessage(), e);
-      } catch (AccumuloSecurityException e) {
-        log.error("{}", e.getMessage(), e);
+      } catch (AccumuloException | TableNotFoundException | AccumuloSecurityException e) {
+        String extentMsg = extent == null ? "" : " for extent: " + extent;
+        log.error("Failed to write metadata updates {} {}", extentMsg, Mutation.prettyPrint(m), e);
       } catch (ConstraintViolationException e) {
-        log.error("{}", e.getMessage(), e);
+        String extentMsg = extent == null ? "" : " for extent: " + extent;
+        log.error("Failed to write metadata updates {} {}", extentMsg, Mutation.prettyPrint(m), e);
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/util/MetadataTableUtil.java:166]: This code is duplicated. It could be moved to a private static method.

File: server/base/src/main/java/org/apache/accumulo/server/util/MetadataTableUtil.java [hunk 5]
```diff
@@ -148,22 +148,24 @@ public static void update(ClientContext context, ZooLock zooLock, Mutation m, Ke
   }
 
   public static void update(Writer t, ZooLock zooLock, Mutation m) {
+    update(t, zooLock, m, null);
+  }
+
+  public static void update(Writer t, ZooLock zooLock, Mutation m, KeyExtent extent) {
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/util/MetadataTableUtil.java:154]: I see the new method here, but I don't see any code that actually uses it, where KeyExtent is non-null.

Comment (jschmidt10) [reply server/base/src/main/java/org/apache/accumulo/server/util/MetadataTableUtil.java]: Good catch, I missed an update.

File: server/base/src/main/java/org/apache/accumulo/server/util/MetadataTableUtil.java [hunk 6]
```diff
@@ -144,27 +144,37 @@ private static void update(ServerContext context, Mutation m, KeyExtent extent)
...
         // retrying when a CVE occurs is probably futile and can cause problems, see ACCUMULO-3096
         throw new RuntimeException(e);
       }
       sleepUninterruptibly(1, TimeUnit.SECONDS);
     }
   }
 
+  private static void logUpdateFailure(Mutation m, KeyExtent extent, Exception e) {
+    String extentMsg = extent == null ? "" : " for extent: " + extent;
+    log.error("Failed to write metadata updates {} {}", extentMsg, m.prettyPrint(), e);
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/util/MetadataTableUtil.java:175]: Removes the second space after the word `updates`:

```suggestion
    String extentMsg = extent == null ? "" : " for extent: " + extent;
    log.error("Failed to write metadata updates{} {}", extentMsg, m.prettyPrint(), e);
```

File: server/base/src/main/java/org/apache/accumulo/server/util/MetadataTableUtil.java [hunk 7]
```diff
@@ -144,27 +144,37 @@ private static void update(ServerContext context, Mutation m, KeyExtent extent)
...
-    update(context, t, zooLock, m);
+    update(context, t, zooLock, m, extent);
   }
 
   public static void update(ServerContext context, Writer t, ServiceLock zooLock, Mutation m) {
+    update(context, t, zooLock, m, null);
+  }
+
+  public static void update(ServerContext context, Writer t, ServiceLock zooLock, Mutation m,
+      KeyExtent extent) {
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/util/MetadataTableUtil.java:155]: I see there's only ever two places in the code that called the old method without the extent. The first is the one you changed above to call the new method, and where extent is never null. The second is in MetaConstraintRetryIT, which also has an extent that is non-null. I think there's no reason to have both update methods. You can just add extent to the existing one, without preserving the old behavior. You only have to update two references, and there's never a need to pass null into the method for extent, or to check for null in the log message where you have the ternary operator to change the method if it's null. If, for some reason it does end up being null, it's not a problem for it to say "for extent null".

File: server/base/src/main/java/org/apache/accumulo/server/util/MetadataTableUtil.java (hunk 6 above)
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/util/MetadataTableUtil.java:175]: This suggestion is superseded by my subsequent suggestion to remove this ternary logic, since extent never needs to be null.

File: server/base/src/main/java/org/apache/accumulo/server/util/MetadataTableUtil.java (hunk 7 above)
Comment (jschmidt10) [reply server/base/src/main/java/org/apache/accumulo/server/util/MetadataTableUtil.java]: Cool. Fixed.
